src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1046 @@
+
+/*---------------------------------------------------------------
+  lsrkStep_AccessARKODEStepMem:
+
+  Shortcut routine to unpack both ark_mem and step_mem structures
+  from void* pointer.  If either is missing it returns ARK_MEM_NULL.
+  ---------------------------------------------------------------*/
+int lsrkStep_AccessARKODEStepMem(void* arkode_mem, const char* fname,
+                                 ARKodeMem* ark_mem, ARKodeLSRKStepMem* step_mem)
+{
+  /* access ARKodeMem structure */
+  if (arkode_mem == NULL)
+  {
+    arkProcessError(NULL, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_ARK_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  *ark_mem = (ARKodeMem)arkode_mem;
+
+  /* access ARKodeLSRKStepMem structure */
+  if ((*ark_mem)->step_mem == NULL)
+  {
+    arkProcessError(*ark_mem, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_LSRKSTEP_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  *step_mem = (ARKodeLSRKStepMem)(*ark_mem)->step_mem;
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_AccessStepMem:
+
+  Shortcut routine to unpack the step_mem structure from
+  ark_mem.  If missing it returns ARK_MEM_NULL.
+  ---------------------------------------------------------------*/
+int lsrkStep_AccessStepMem(ARKodeMem ark_mem, const char* fname,
+                           ARKodeLSRKStepMem* step_mem)
+{
+  /* access ARKodeLSRKStepMem structure */
+  if (ark_mem->step_mem == NULL)
+  {
+    arkProcessError(ark_mem, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_LSRKSTEP_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  *step_mem = (ARKodeLSRKStepMem)ark_mem->step_mem;
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_CheckNVector:
+
+  This routine checks if all required vector operations are
+  present.  If any of them is missing it returns SUNFALSE.
+  ---------------------------------------------------------------*/
+sunbooleantype lsrkStep_CheckNVector(N_Vector tmpl)
+{
+  if ((tmpl->ops->nvclone == NULL) || (tmpl->ops->nvdestroy == NULL) ||
+      (tmpl->ops->nvlinearsum == NULL) || (tmpl->ops->nvconst == NULL) ||
+      (tmpl->ops->nvscale == NULL) || (tmpl->ops->nvwrmsnorm == NULL))
+  {
+    return (SUNFALSE);
+  }
+  return (SUNTRUE);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_SprRadUpdateLogic:
+
+  This routine checks if the step is accepted or not and reassigns
+  the SprRad update flags accordingly.
+  ---------------------------------------------------------------*/
+
+int lsrkStep_ComputeNewSprRad(ARKodeMem ark_mem, ARKodeLSRKStepMem step_mem)
+{
+  int retval = SUN_SUCCESS;
+
+  if ((step_mem->isextspr))
+  {
+    retval = step_mem->extspr(ark_mem->tn, &step_mem->sprad, ark_mem->user_data);
+    step_mem->sprad *= step_mem->sprsfty;
+    step_mem->sprad = SUNRabs(step_mem->sprad);
```

Comment:
I recommend that we update the interface to `extspr` now so that it will work when we eventually support problems with complex eigenvalues.  Currently, `extspr` is hard-coded to receive a single `sunrealtype` value, so it obviously cannot work in the complex case.